### PR TITLE
kip-146: Random proposer selection based on RANDAO

### PR DESCRIPTION
**Pull request by blukat29**
## Proposed changes

Implement KIP-146 random proposer selection based on `header.MixHash`

- Store MixHash in IstanbulSnapshot
- Call SetMixHash in `snapshot.apply()` alongside SetBlockNum
- Modify proposer selection algorithm `weightedRandomProposer`
- Modify committee selection algorithm `SubListWithProposer`
- Fix tests bug regarding StakingManager (only triggered when ProposerPolicy=WeightedRandom)

## Types of changes

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: consensus/istanbul/validator/weighted.go [hunk 1]
```diff
@@ -359,6 +377,19 @@ func (valSet *weightedCouncil) SubListWithProposer(prevHash common.Hash, propose
...
+	rules := fork.Rules(new(big.Int).SetUint64(valSet.blockNum + 1))
+	// After Randao: SelectRandaoCommittee
+	if rules.IsRandao {
+		// This committee must include proposers for all rounds because
+		// the proposer is picked from the this committee. See weightedRandomProposer().
+		if valSet.mixHash == nil {
+			logger.Error("no mixHash", "number", valSet.blockNum)
+			return nil
+		}
+		return SelectRandaoCommittee(validators, committeeSize, valSet.mixHash)
```
Comment (hyeonLewis) [inline consensus/istanbul/validator/weighted.go:389]: Don't we need the following nil check?
```suggestion
		if committee := SelectRandaoCommittee(validators, committeeSize, valSet.mixHash); committee != nil {
		    return committee
		}
		return validator 
```

**Review by hyeonLewis — COMMENTED** [review]
Will tests for the randao proposer selection be added later?

**Comment by blukat29** [comment]
@hyeonLewis I'll add unit tests in weighted_random_test.go.

**Comment by blukat29** [comment]
@hyeonLewis Tests added. PTAL.

**Review by hyeonLewis — APPROVED** [review]

File: consensus/istanbul/validator/weighted.go [hunk 2]
```diff
@@ -359,6 +375,22 @@ func (valSet *weightedCouncil) SubListWithProposer(prevHash common.Hash, propose
 		return []istanbul.Validator{proposer}
 	}
 
+	rules := fork.Rules(new(big.Int).SetUint64(valSet.blockNum + 1))
```
Comment (ian0371) [inline consensus/istanbul/validator/weighted.go:378]: Is it always true that `valSet.blockNum + 1 == view.Sequence`?
Can there be a case where `view.Sequence` is Randao but `valStt.blockNum + 1` is not?

Comment (blukat29) [reply consensus/istanbul/validator/weighted.go]: Fixed to use `fork.Rules(view.Sequence).IsRandao` instead, just like L418 `fork.Rules(view.Sequence).IsIstanbul`.

There is another usage of `fork.Rules(...blockNum+1...)` at `weightedRandomProposer()` selector.
Here, I empirically verified with `TestRandao_Deploy` that `valSet.blockNum + 1` is the currently created block.

**Review by ian0371 — APPROVED** [review]

**Review by hyeonLewis — APPROVED** [review]

**Comment by hyunsooda** [comment]
@blukat29, Before the randao fork, it is imperative to deploy both the registry and KIP113 contracts. Additionally, every existing validator must register their key with the BLS contract. I have validated this PR by adjusting the `genesis.conf` file to configure the necessary storage states. It's worth noting that the conventional hard fork methodologies do not incorporate this particular patch. How can each CN ensure the proper execution of these essential setup steps?

**Review by hyunsooda — APPROVED** [review]

**Comment by blukat29** [comment]
@hyunsooda
For testing purposes, you can use homi to start the chain with Registry and KIP-113 filled up (#2028).

In Baobab, the upgrade order would be:
1. Deploy KIP-113 by sending a transaction
2. Setup `ChainConfig.RandaoNumber` and `ChainConfig.RandaoRegistry` with the KIP-113 address. Release it as v1.12.0.
3. CNs upgrade to v1.12.0 and prepare their BlsPublicKeyInfo using `kcn account bls-info`.
4. Populate KIP-113 by sending transactions
5. Pass the hardfork block.

The Cypress upgrade order will be similar but with v1.12.1.

**Comment by blukat29** [comment]
@yoomee1313 PTAL.

**Review by yoomee1313 — APPROVED** [review]

**Pull request merged**